precepte-core-cats/src/test/scala/PrecepteSpec.scala

```diff
@@ -532,39 +577,104 @@ class PrecepteSpec extends FlatSpec with ScalaFutures with Inside {
         _ <- Precepte(tags("sub"))(p4)
         _ <- p5
         _ <- Precepte(tags("sub2"))(for {
-            _ <- (p1, p2, p3, p4).tupled
-            _ <- p6
-            _ <- (p4, p5, Precepte(tags("sub3"))(p6)).tupled
-            _ <- p7
-          } yield ())
+          _ <- (p1, p2, p3, p4).tupled
+          _ <- p6
+          _ <- (p4, p5, Precepte(tags("sub3"))(p6)).tupled
+          _ <- p7
+        } yield ())
       } yield ()
 
     import cats.~>
 
-    type SF[T] = (ST[Int], Future[T])
-
-    object Mon extends (SF ~> Future) {
-      def apply[A](f: SF[A]): Future[A] = {
-        val t0 = System.nanoTime()
-        val path = f._1.managed.path
-        val method = path.last.tags.callee.value
-        f._2.map { a =>
-          val t1 = System.nanoTime()
-          val duration = t1 - t0
-          // TODO: Store measure and test result
-          println(s"$method $duration")
-          a
-        }
-      }
+    type P[X] = Precepte[BaseTags, MS, Int, Future, X]
+    object P extends Precepte.API[BaseTags, MS, Int, Future]
+
+    object Mon extends (P ~> P) {
+      def apply[A](f: P[A]): P[A] =
+        for {
+          state0 <- P.get
+          t0 <- P.delay(System.nanoTime())
+          a <- f
+          t1 <- P.delay(System.nanoTime())
+          _ <- P.delay {
+
+            val path = state0.managed.path
+            val method = path.last.tags.callee.value
+            val duration = t1 - t0
+
+            // TODO: Store measure and test result
+            println(s"$method $duration")
+          }
+        } yield a
     }
 
     implicit val intSG = new cats.Semigroup[Int] {
       def combine(f1: Int, f2: Int) = f1 + f2
     }
     def nostate = ST(Span.gen, env, Vector.empty, 0)
 
-    val res = CatsExt(p8).mapSuspension(Mon).eval(nostate).futureValue
+    val res = new CatsExt(p8).mapSuspension(Mon).eval(nostate).futureValue
     res should ===(())
+
+  }
+
+  it should "catch exceptions in Future" in {
+    object P extends Precepte.API[BaseTags, default.MS, Unit, Future]
+
+    def f(i: Long): P.precepte[Int] = {
+      val m =
+        for {
+          v3 <- P.deferredLift(Future(3 / (i % 3)))
+          v5 <- P.deferredLift(Future(5 / (i % 5)))
+        } yield v3 + v5
+
+      m.map(_ => 0).recoverWith { _ =>
+        P.pure(1)
+      }
+    }
+
+    def g(i: Long): P.precepte[Int] =
+      if (i <= 0) P.pure(0)
+      else
+        for {
+          x <- f(i)
+          q <- g(i - 1)
+        } yield x + q
+
+    val n = 150L
+    val exceptions =
```

Comment:
Maybe add a comment saying "numbers divisible by either 3 or 5 will raise a DivisionByZero in f(), meaning they'll count as '1' (see the `recoverWith`). We expect `g()`'s return value to be the same as the number of thrown exceptions"